src/dbnode/storage/database.go

```diff
@@ -317,14 +309,61 @@ func (d *db) UpdateOwnedNamespaces(newNamespaces namespace.Map) error {
 			"restart the process if you want changes to take effect")
 	}
 
-	// enqueue bootstraps if new namespaces
 	if len(adds) > 0 {
-		d.queueBootstrapWithLock()
-	}
+		if noBootstrapNeeded {
+			d.Lock()
+			defer d.Unlock()
+			return d.addNamespacesWithLock(adds)
+		}
+
+		if mediator := d.mediator; mediator == nil || !mediator.IsOpen() {
+			d.log.Info("mediator is closed, adding namespaces")
+			d.Lock()
+			if err := d.addNamespacesWithLock(adds); err != nil {
+				d.Unlock()
+				d.log.Error("unable to add namespaces", zap.Error(err))
+				return err
+			}
+			d.Unlock()
 
+			d.enqueueBootstrap()
```

Comment:
This can be written without multiple unlock statements:
```suggestion
			d.Lock()
			err := d.addNamespacesWithLock(adds)
			d.Unlock()
			if err != nil {
				d.log.Error("unable to add namespaces", zap.Error(err))
				return err
			}

			d.enqueueBootstrap()
```